### PR TITLE
Add compiler negative tests for resource access action

### LINKED ISSUE #36637: Add compiler negative tests for resource access action
**Description:**

Add compiler negative tests for resource access action and resource method declarations with resource path parameters containing only the type name

### PR DESCRIPTION
**Pull request by SasinduDilshara**
Fixes #36637

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/services/ClientServiceDeclTest.java [hunk 1]
```diff
@@ -0,0 +1,365 @@
+package org.ballerinalang.test.services;
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/services/ClientServiceDeclTest.java:1]: Shall we add the copyright headers?

Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/services/ClientServiceDeclTest.java:1]: I will add

Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/services/ClientServiceDeclTest.java:19]: Updated. e6adb7809bcf24f0faff2f7484b560a29cd03c36

File: tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_call_negative.bal [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            return 1;
+        }
+
+        resource function get path/[int a]/foo2(string name, string address) returns int {
+            return 1;
+        }
+    };
+
+    int result = successClient->/path/[10*11]/foo(123);
+    result = successClient->/path/[1]/foo("x", "y"); // No Error occurred
```
Comment (KavinduZoysa) [inline tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_call_negative.bal:38]: @rdulmina, why is not this an error?

Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_call_negative.bal:38]: Created a issue for this. https://github.com/ballerina-platform/ballerina-lang/issues/36687

File: tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_call_negative.bal [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    result = successClient->/path/[1]/foo("x", "y", "z");
```
Comment (KavinduZoysa) [inline tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_call_negative.bal:39]: @rdulmina, we are getting `too many arguments in call to 'get()'`
But there are few get methods. Can't we specify what is the exact method?
Like,
`too many arguments in call to 'get path/[int a]/foo2(string name, string address)'`

File: tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_declaration_negative.bal [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        resource function get errorPath/[error]() returns int {
+            return 1;
+        }
+
+        resource function get errorPath2/[error a]() returns int {
+            return 1;
+        }
+
+    };
+}
```
Comment (KavinduZoysa) [inline tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_declaration_negative.bal:78]: @SasinduDilshara, Shall we add few tests with valid/invalid module-type-defn?

Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_declaration_negative.bal:78]: Also we need to a tests for invalid returns

Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_declaration_negative.bal:78]: We have invalid return tests in `client_resource_params_negative.bal`.
Do we need add more?

File: tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_call_negative.bal (hunk 3 above)
Comment (rdulmina) [reply tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_call_negative.bal:39]: Will have a look

File: tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_declaration_negative.bal (hunk 4 above)
Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_declaration_negative.bal:78]: I will add tests for `valid/invalid module-type-defn`

File: tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_call_negative.bal (hunk 2 above)
Comment (rdulmina) [reply tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_call_negative.bal:38]: Fixed with [c6b8d7b](https://github.com/ballerina-platform/ballerina-lang/pull/36600/commits/c6b8d7beea75ac61238627c7b624be6cdaca6a40)

File: tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_call_negative.bal (hunk 3 above)
Comment (rdulmina) [reply tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_call_negative.bal:39]: Created issue to address this later https://github.com/ballerina-platform/ballerina-lang/issues/36709

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 5]
```diff
@@ -2513,6 +2513,27 @@ public void visit(BLangInvocation.BLangActionInvocation actionInvocation, Analyz
...
+    public void visit(BLangInvocation.BLangResourceAccessInvocation resourceActionInvocation, AnalyzerData data) {
+        validateInvocationInMatchGuard(resourceActionInvocation);
+        analyzeExpr(resourceActionInvocation.expr, data);
+        analyzeExprs(resourceActionInvocation.requiredArgs, data);
+        analyzeExprs(resourceActionInvocation.restArgs, data);
+        analyzeExpr(resourceActionInvocation.resourceAccessPathSegments, data);
+        resourceActionInvocation.invokedInsideTransaction = data.withinTransactionScope;
+        
+        if (resourceActionInvocation.flagSet.contains(Flag.TRANSACTIONAL) && !data.withinTransactionScope) {
+            dlog.error(resourceActionInvocation.pos, DiagnosticErrorCode.TRANSACTIONAL_FUNC_INVOKE_PROHIBITED);
```
[the remark this answers is not in the document]
Comment (SasinduDilshara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java:2527]: I will add

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 6]
```diff
@@ -3500,6 +3503,105 @@ public void visit(BLangInvocation.BLangActionInvocation aInv, AnalyzerData data)
...
+        
+        // Filter the resource methods in the list by resource access method name
+        resourceFunctions.removeIf(func -> !func.accessor.value.equals(resourceAccessInvocation.name.value));
+        
+        if (resourceFunctions.size() == 0) {
+            dlog.error(resourceAccessInvocation.pos, 
+                    DiagnosticErrorCode.UNDEFINED_RESOURCE_METHOD, resourceAccessInvocation.name, lhsExprType);
+            data.resultType = symTable.semanticError;
+        } else if (resourceFunctions.size() > 1) {
+            dlog.error(resourceAccessInvocation.pos, DiagnosticErrorCode.AMBIGUOUS_RESOURCE_ACCESS_NOT_YET_SUPPORTED);
```
[the remark this answers is not in the document]
Comment (SasinduDilshara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:3553]: I will add

[the remark this answers is not in the document]
Comment (SasinduDilshara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java]: Added. b731811e1ec16fc62289fbdaff5f454fd9dd283c

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (SasinduDilshara) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: Added. b731811e1ec16fc62289fbdaff5f454fd9dd283c

File: tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_declaration_negative.bal (hunk 4 above)
Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_declaration_negative.bal:78]: Added. b731811e1ec16fc62289fbdaff5f454fd9dd283c

File: tests/jballerina-unit-test/src/test/resources/test-src/services/client_transactional_resource_negative.bal [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+public function testTransactionCall() {
+    var successClient = client object {
+        transactional resource function get path/[string]() returns int {
+            return 1;
+        }
+        resource function get path2/[string]() {
+            testFunction();
+        }
+    };
+    int result = successClient->/path/["a"]; // No Error Found. Issue - #36736.
```
Comment (rdulmina) [inline tests/jballerina-unit-test/src/test/resources/test-src/services/client_transactional_resource_negative.bal:26]:
```suggestion
    int _ = successClient->/path/["a"]; // No Error Found. Issue - #36736.
```

let's use `_` for unused variables

File: tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_declaration_negative.bal [hunk 8]
```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2022 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (rdulmina) [inline tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_declaration_negative.bal:1]: We already have a test file `resource_function_with_invalid_path_param_type_negative.bal` which includes these kind of tests. Let's add these tests to that

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/services/ClientServiceDeclTest.java [hunk 9]
```diff
@@ -0,0 +1,421 @@
+/*
```
Comment (rdulmina) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/services/ClientServiceDeclTest.java:1]: rename to `ClientResourceAccessActionNegativeTest` and move this to action folder. move tests files accordingly

File: tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_declaration_negative.bal (hunk 8 above)
Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/resources/test-src/services/client_resource_declaration_negative.bal]: Updated 4bc5561b9d0ff49d9cbd3499068a7004712ef736

**Review by KavinduZoysa — APPROVED** [review]

File: tests/jballerina-unit-test/src/test/resources/test-src/action/client_resource_params_negative.bal [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    xml _ = successClient->/intParamPath(2);
+    xml _ = successClient->/booleanParamPath(true);
+    xml _ = successClient->/floatParamPath(2.3f);
+    xml _ = successClient->/decimalPath(3.2d);
+
+    string _ = successClient->/stringParamPath(2);
+    int _ = successClient->/intParamPath("string");
+    boolean _ = successClient->/booleanParamPath(3.4f);
+    float _ = successClient->/floatParamPath(2.3d);
+    decimal _ = successClient->/decimalPath(true);
```
Comment (rdulmina) [inline tests/jballerina-unit-test/src/test/resources/test-src/action/client_resource_params_negative.bal:80]: We have already added these tests in `client_resource_call_negative.bal` right?. I think we can rename this file to `client_resource_access_return_type_negative_test.bal`

File: tests/jballerina-unit-test/src/test/resources/test-src/action/client_resource_path_negative.bal [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    int _ = successClient->/intPath/'5;
+    int _ = successClient->/intPath/foo/foo2/bar/bar2/foo3;
+    int _ = successClient->/path/[intNilArgument];
+    int _ = successClient->/path/[nilArgument];
+    int _ = successClient->/path/[getIntOrNilArgument()];
+}
+
+function getIntOrNilArgument() returns int? {
+    return intOrNilArgument;
+}
```
Comment (rdulmina) [inline tests/jballerina-unit-test/src/test/resources/test-src/action/client_resource_path_negative.bal:95]: We need to add negative tests to validate the computed resource access segment type and rest segment type in client-resource-access-expression

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/action/ClientResourceAccessActionNegativeTest.java [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+
+import org.ballerinalang.test.BAssertUtil;
+import org.ballerinalang.test.BCompileUtil;
+import org.ballerinalang.test.CompileResult;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+/**
+ * Tests calling of a client resource access action.
```
Comment (rdulmina) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/action/ClientResourceAccessActionNegativeTest.java:29]:
```suggestion
 * Tests for invalid client resource access action.
```

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/action/ClientResourceAccessActionNegativeTest.java [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+ */
+public class ClientResourceAccessActionNegativeTest {
+    CompileResult clientResourceCallNegative, clientResourcePathNegative,
```
[the remark this answers is not in the document]
Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/action/ClientResourceAccessActionNegativeTest.java]: Updated 954c43cfd50ecab237cda237b586ee8324855148

File: tests/jballerina-unit-test/src/test/resources/test-src/action/client_resource_call_negative.bal [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+() nilArgument = ();
+
+CustomString customString = "x";
+CustomString2 customString2 = customString;
+
+const int intConstant = 1;
+var varInt = 1;
+var varBoolean = true;
+
+public function testClientResourceFunctionCallError() {
```
[the remark this answers is not in the document]
Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/resources/test-src/action/client_resource_call_negative.bal]: Updated 954c43cfd50ecab237cda237b586ee8324855148

File: tests/jballerina-unit-test/src/test/resources/test-src/action/client_resource_call_negative.bal [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+    int _ = successClient->/path/[1]/foo2(address = 23, "name");
+    int _ = successClient->/path.post("a");
+    int _ = successClient->/path/[1]/foo2(name = customString, 23);
+    int _ = successClient->/path/[1]/foo(customString, name = 23);
+    int _ = successClient->/path/[1]/foo(customString, customString);
+    int _ = successClient->/path/[1]/foo(customString, customString, customString);
+    int _ = successClient->/path.post(customString);
+}
+
+public function testClientResourceFunctionCallAmbiguousError() {
```
[the remark this answers is not in the document]
Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/resources/test-src/action/client_resource_call_negative.bal]: Updated 954c43cfd50ecab237cda237b586ee8324855148

File: tests/jballerina-unit-test/src/test/resources/test-src/action/client_resource_params_negative.bal (hunk 10 above)
Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/resources/test-src/action/client_resource_params_negative.bal]: Updated 954c43cfd50ecab237cda237b586ee8324855148

File: tests/jballerina-unit-test/src/test/resources/test-src/action/client_resource_call_negative.bal [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+        }
+
+        resource function get foo2/path() returns int {
+            return 1;
+        }
+    };
+
+    int _ = successClient->/foo/path();
+    int _ = successClient->/foo2/path();
+    int _ = successClient->/foo2/["path"]();
```
[the remark this answers is not in the document]
Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/resources/test-src/action/client_resource_call_negative.bal:147]: Added b07db9bcff0445bb32760089169ec8ed48adf2a4

File: tests/jballerina-unit-test/src/test/resources/test-src/action/client_resource_path_negative.bal (hunk 11 above)
Comment (SasinduDilshara) [reply tests/jballerina-unit-test/src/test/resources/test-src/action/client_resource_path_negative.bal:157]: Added b07db9bcff0445bb32760089169ec8ed48adf2a4

**Review by rdulmina — APPROVED** [review]

**Review by gimantha — APPROVED** [review]

**Pull request merged**